{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "from plotly.offline import iplot, init_notebook_mode\n",
    "\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_colwidth', 0)\n",
    "\n",
    "import plotly.plotly as py\n",
    "import plotly.graph_objs as go\n",
    "from ipywidgets import HTML\n",
    "from ipywidgets import HBox, VBox\n",
    "\n",
    "\n",
    "############################# put your antismash 5 gbk output in here #############################\n",
    "\n",
    "#DATA import\n",
    "gb = 'Azum.gbk'\n",
    "\n",
    "################################################################################################\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##Lists and dictionaries \n",
    "\n",
    "##Parse gbk to lists\n",
    "locos = []\n",
    "for record in SeqIO.parse(gb, \"gb\"):\n",
    "     for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            locos.append(feature.location)\n",
    "            \n",
    "\n",
    "\n",
    "            \n",
    "#for mapping tick marks \n",
    "x = [x for x in range(0,locos[-1].end, 1000)]+[locos[-1].end]\n",
    "\n",
    "\n",
    "\n",
    "##For mapping gene type colors \n",
    "gene_colors = {'biosynthetic': 'rgb(126, 16, 25)',\n",
    "               'regulatory': 'rgb(50,138,89)',\n",
    "               'biosynthetic-additional':'rgb(238,110,119)' ,\n",
    "               'resistance': 'rgb(235,47,235)',\n",
    "               'transport': 'rgb(102, 151, 234)',\n",
    "               'N':'rgb(128,128,128)',\n",
    "               'other':'rgb(101,202,56)'\n",
    "}\n",
    "\n",
    "\n",
    "##For mapping points to orfs  \n",
    "orf_points = [(location.start + location.end)/2 for location in locos]\n",
    "\n",
    "\n",
    "\n",
    "all_features = []\n",
    "for record in SeqIO.parse(gb, \"gb\"):\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            all_features.append(feature)\n",
    "            \n",
    "\n",
    "names = [feature.qualifiers.get('gene')[0] for feature in all_features if feature.qualifiers.get('gene')]\n",
    "if not names:\n",
    "    all_locus_tags = [feature.qualifiers.get('locus_tag')[0] for feature in all_features if feature.qualifiers.get('locus_tag')] \n",
    "    for tag in all_locus_tags:\n",
    "        if tag not in names:\n",
    "            names.append(tag)\n",
    "            \n",
    "p_strings = [feature.qualifiers['translation'][0] for feature in all_features]\n",
    "\n",
    "#write the seqs to a multifasta file temp.faa\n",
    "for_fasta = [SeqRecord(Seq(seq_string, IUPAC.protein), id=seq_id) for seq_string, seq_id  in zip(p_strings, names)]\n",
    "SeqIO.write(for_fasta, f\"{gb}.faa\", 'fasta')\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now you need to do the BLASTp on the *.gbk.faa file, that was just generated, using the webserver and DL the xml to this directory and run the next cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "94e52bcc2ed048b9a7c1adae51ca14fa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(FigureWidget({\n",
       "    'data': [{'hoverinfo': 'text',\n",
       "              'marker': {'size': 1},\n",
       "        â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "blast_xml = f\"{gb}.xml\"\n",
    "\n",
    "blast_records = NCBIXML.parse(open(blast_xml))\n",
    "data_dict = {}\n",
    "\n",
    "for blast_record in blast_records:\n",
    "    \n",
    "    data = []\n",
    "    for alignment in blast_record.alignments: \n",
    "        data.append(alignment.title.split('|')[2].replace('>gb',''))\n",
    "        \n",
    "        for hsp in alignment.hsps:            \n",
    "            data.append(round(((hsp.identities/hsp.align_length)*100),2))\n",
    "            data_dict[blast_record.query]=data\n",
    "            break\n",
    "        \n",
    "        break\n",
    "         \n",
    "\n",
    "            \n",
    "gene_types = [feature.qualifiers.get('gene_kind', [None])[0] for feature in all_features if feature.type == 'CDS']\n",
    "            \n",
    "blast_table = pd.DataFrame.from_dict(data_dict, orient='index',columns=['Top Blast Hit','% identity'])   \n",
    "blast_table['gene_type'] = gene_types\n",
    "\n",
    "\n",
    "\n",
    "##Functions\n",
    "##To get gene colors, needs to be run through AS v5. \n",
    "def gene_type_color(feature):\n",
    "    gene_type = feature.qualifiers.get('gene_kind', 'None')\n",
    "    \n",
    "    return gene_colors[gene_type[0]]\n",
    "\n",
    "\n",
    "def frd_to_svg(feature):\n",
    "    \n",
    "    if len(feature.location) < locos[-1].end/100:\n",
    "        nose_offset = len(feature.location)\n",
    "    else:\n",
    "        nose_offset = locos[-1].end/100\n",
    "    \n",
    "    \n",
    "    start = feature.location.start\n",
    "    length = (feature.location.start + len(feature.location)) - nose_offset\n",
    "    nose = length + nose_offset \n",
    "    \n",
    "    \n",
    "    return f'M {start},0.1 L{start},1.1 L{length},1.1 L{nose},.6 L{length},0.1 L {start},0.1'\n",
    "\n",
    "\n",
    "def rev_to_svg(feature):\n",
    "    \n",
    "    if len(feature.location) < locos[-1].end/100:\n",
    "        nose_offset = len(feature.location)\n",
    "    else:\n",
    "        nose_offset = locos[-1].end/100\n",
    "    \n",
    "    \n",
    "    start = feature.location.start + nose_offset \n",
    "    length = (feature.location.start + len(feature.location)) \n",
    "    nose = start - nose_offset \n",
    "    \n",
    "    \n",
    "    return f'M {start},0.1 L{length},0.1 L{length},1.1 L{start},1.1 L{nose},0.6 L {start},0.1'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "plot_range = locos[-1].end ## total length of plot to show (gene cluster length plus 5%) \n",
    "\n",
    "##List of all the + strand gens to show on the plot\n",
    "            \n",
    "frd_orfs = [{'type': 'path',\n",
    "            'path': frd_to_svg(feature),\n",
    "            \n",
    "           'fillcolor': gene_type_color(feature), #make funct for fill color (fill_col(feature_location))\n",
    "            \n",
    "           'line': {'color': gene_type_color(feature),\n",
    "                    'width': 1.2}\n",
    "                    } \n",
    "             \n",
    "            for feature in all_features if feature.location.strand == 1]   \n",
    "\n",
    "rev_orfs = [{'type': 'path',\n",
    "            'path': rev_to_svg(feature),\n",
    "            \n",
    "           'fillcolor': gene_type_color(feature), #make funct for fill color (fill_col(feature_location))\n",
    "            \n",
    "           'line': {'color': gene_type_color(feature),\n",
    "                   'width': 1.2}\n",
    "                    } \n",
    "             \n",
    "            for feature in all_features if feature.location.strand == -1]   \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "trace = go.Scatter(\n",
    "    x = orf_points,\n",
    "    y = [.5]*len(orf_points),\n",
    "    mode = 'markers',\n",
    "    hoverinfo = 'text',\n",
    "    text = names\n",
    ")\n",
    "\n",
    "\n",
    "data = [trace]\n",
    "\n",
    "layout={'xaxis': {\n",
    "        'showgrid':False,\n",
    "        'zeroline':False,\n",
    "        'showline':False,\n",
    "        'tickmode':'linear',\n",
    "        'ticks':'',\n",
    "        'showticklabels':False,\n",
    "        'range': [0 , plot_range ]},\n",
    "        \n",
    "        'yaxis':{\n",
    "        'range': [-4 ,4],\n",
    "        'showgrid':False,\n",
    "        'zeroline':True,\n",
    "        'showline':False,\n",
    "        'ticks':'',\n",
    "        'showticklabels':False},\n",
    "         \n",
    "        'shapes': frd_orfs + rev_orfs\n",
    "       }\n",
    "\n",
    "\n",
    "fig = go.FigureWidget(data,layout)\n",
    "\n",
    "\n",
    "#updates to figures if using ipywidgets\n",
    "\n",
    "fig.layout.title = 'Gene cluster'\n",
    "\n",
    "scatter = fig.data[0]\n",
    "scatter.marker.size = 1\n",
    "\n",
    "details = HTML()\n",
    "\n",
    "def hov_func(trace, points, state):\n",
    "    ind = points.point_inds[0]\n",
    "    details.value = blast_table.iloc[ind].to_frame().to_html()\n",
    "\n",
    "scatter.on_hover(hov_func)\n",
    "\n",
    "VBox([fig, HBox([details])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
